lib/eco/treemanager.py

```diff
@@ -2191,16 +2191,17 @@ def lbox_expand_test(self, lbox):
         op = self.get_parser(outer_root)
         ol = self.get_lexer(outer_root)
         r2 = IncrementalRecognizer(op.syntaxtable, ol.lexer, outer_lang, None)
-        r2.mode_limit_tokens_new = r.mode_limit_tokens_new
+        r2.mode_limit_tokens_new = False
         r2.preparse(outer_root, lbox)
+        r2.parse_after(lbox)
         filtered = []
-        start = lbox
         for e in r.possible_ends:
             if e.lookup == "<ws>" or e.lookup == "<return>":
                 continue
-            if r.parse_until(start, e):
+            tmp = list(r2.state)
+            if r2.parse_after(e.next_term):
                 filtered.append(e)
-            start = e
+            r2.state = tmp
```

Comment:
Why are we (in effect) doing `r2.state = list(r2.state)`? What's the `list` call for?

Reply:
`state` is the (admittedly poorly named) parsing stack of the recogniser. After initialising the recogniser with `preparse`, we then want to try and check if each element in `possible_end` can be parsed in that state, so we have to reset the parse stack back after every iteration.

Reply:
Yes, but it's the `list` bit that I'm confused by.

Reply:
It just creates a copy of the parse stack which we can then later apply to r2, to reset the parse stack back to what it was before.

Reply:
I can also use `r.state[:]` if you prefer that.

Reply:
Oh, I see, you're doing a shallow copy! Ah. Yes, `[:]` will make this clearer. I thought you were converting from a non-list to a list!

Reply:
Fixed in the latest commit.